src/Common.OData.ApiExplorer/AspNet.OData/DefaultModelTypeBuilder.cs

```diff
@@ -66,6 +67,14 @@ public Type NewStructuredType( IEdmStructuredType structuredType, Type clrType,
                 var propertyType = property.PropertyType;
                 var structuredTypeRef = structuralProperty.Type;
 
+                if ( passedTypes.TryGetValue( structuredTypeRef.ToString(), out var passedType ) )
+                {
+                    properties.Add(new ClassProperty(property, passedType));
+                    continue;
+                }
+
+                passedTypes.Add(structuredTypeRef.ToString(), propertyType);
```

Comment:
Adding the visited type needs to be added toward the end; probably after line 102. The `propertyType` variable can be mutated after this line, which means the wrong type will be stored in the dictionary.